pydantic/json_schema.py

```diff
@@ -125,6 +125,34 @@ class PydanticJsonSchemaWarning(UserWarning):
 CoreModeRef = tuple[CoreRef, JsonSchemaMode]
 JsonSchemaKeyT = TypeVar('JsonSchemaKeyT', bound=Hashable)
 
+# ##### Regex for Decimal JSON Schema Generation #####
+
+_DECIMAL_JSON_VALIDATION_MAX_DIGIT_LOOKAHEAD_PATTERN = (
+    r'(?=\d{{0,{max_digits}}}0*$'  # Positive lookahead for max_digits, allows trailing zeroes
```

Comment:
This will match arbitrary trailing zeros, which is only allowed after the `.` separator:

```python
ta = TypeAdapter(Annotated[Decimal, Field(max_digits=4)])

ta.validate_python("30000000") # ValidationError, passes regex:

ta.json_schema()['anyOf'][1]['pattern']
#> '^-?0*(?=\\d{0,4}0*$|(?=.*\\..*)[\\d\\.]{0,5}0*$)\\d{1,}(?:\\.\\d{0,})?0*$'
```

Reply:
But I don't know if this is reasonably easy to support, so perhaps we should use a simpler (but one that accepts invalid inputs in some cases) regex when having only `max_digits` specified.